### PR TITLE
Add NNX surgery guide

### PR DESCRIPTION
**Pull request by IvyZX**
Add a guide for common model surgery scenarios and how to use NNX state and modules to fluently surgery around.

### CODE REVIEW

File: docs/nnx/surgery.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+  new_state = nnx.state(get_model())
+  return nnx.State.merge(new_state, old_state)
+
+good_model = nnx.eval_shape(get_lora_model)
+# Note that `old_state` will be deleted after this `partial_init` call.
+new_state = partial_init(old_state, get_lora_model)
+nnx.update(good_model, new_state)
+
+# They are equivalent. The way RNGs are propagated in the new model is also preserved.
+np.testing.assert_allclose(simple_model(x), good_model(x))
```
Comment (cgarciae) [inline docs/nnx/surgery.md:260]: I'm a bit confused by this section, it seems to be doing the same operation in two different ways. I was thinking we could show case the use of `nnx.jit` here and use something like:

```python
old_state = nnx.state(TwoLayerMLP(4, rngs=nnx.Rngs(0)))

@functools.partial(nnx.jit, donate_argnums=0, static_argnums=1)
def partial_init(old_state, rngs):
  model = TwoLayerMLP(4, rngs=rngs)
  # add existing create
  nnx.update(model, old_state)
  # create new state
  model.linear1 = nnx.LoRALinear(4, 4, lora_rank=3, rngs=rngs))
  return model
  
  model = partial_init(old_state, nnx.Rngs(0))
```

Comment (IvyZX) [reply docs/nnx/surgery.md:283]: Hmm, for some reason the code you proposed didn't really save memory.

I added some lines to explain this section and also showed how to print the memory cost at each step. You can see that the `jax.jit` cases skipped the creation of extra unused params in `LoRALinear`. I don't know why your `nnx.jit` version doesn't repro this...

You can run this piece of code in a standalone python file to verify what I've seen. If `nnx.jit` can do this I'm happy to switch to `nnx.jit` instead.

Comment (cgarciae) [reply docs/nnx/surgery.md:283]: My bad, I put `nnx.update` in the middle but it should be at the end else the `model.linear1 =` assignment overrides the updated state. Here is the fixed code:

```python
  class TwoLayerMLP(nnx.Module):
      def __init__(self, dim, rngs: nnx.Rngs):
        self.linear1 = nnx.Linear(dim, dim, rngs=rngs)
        self.linear2 = nnx.Linear(dim, dim, rngs=rngs)

      def __call__(self, x):
        x = self.linear1(x)
        return self.linear2(x)

    old_state = nnx.state(TwoLayerMLP(4, rngs=nnx.Rngs(0)))

    print(f'Arrays start: {len(jax.live_arrays())}')
    print(len(jax.tree.leaves(old_state)))

    @partial(nnx.jit, donate_argnums=0, static_argnums=1)
    def partial_init(old_state, rngs):
      model = TwoLayerMLP(4, rngs=rngs)
      # create new state
      model.linear1 = nnx.LoRALinear(4, 4, lora_rank=3, rngs=rngs)
      # add existing create
      nnx.update(model, old_state)
      return model

    model = partial_init(old_state, nnx.Rngs(0))

    print(f'Arrays end: {len(jax.live_arrays())}')
    print(len(jax.tree.leaves(nnx.state(model))))
```

Comment (IvyZX) [reply docs/nnx/surgery.md]: Thanks! Added your code and replaced the `jax.jit` code. Thank you!

**Review by cgarciae — APPROVED** [review]

**Pull request merged**